scripts/inject-deterministic.py

```diff
@@ -1,50 +1,171 @@
-from __future__ import absolute_import, print_function
-
+# from __future__ import absolute_import, print_function
+import base64
+import hashlib
 import re
 import time
+
 from mitmproxy import ctx
 
 
 class AddDeterministic():
 
+    def get_csp_directives(self, headers):
+        csp = headers.get("Content-Security-Policy", "")
+        return [d.strip() for d in csp.split(";")]
+
+    def get_csp_script_sources(self, headers):
+        sources = []
+        for directive in self.get_csp_directives(headers):
+            if directive.startswith("script-src "):
+                sources = directive.split()[1:]
+        return sources
+
+    def get_nonce_from_headers(self, headers):
+        """
+        get_nonce_from_headers returns the nonce token from a
+        Content-Security-Policy (CSP) header's script source directive.
+
+        Note:
+        For more background information on CSP and nonce, please refer to
+        https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/
+        Content-Security-Policy/script-src
+        https://developers.google.com/web/fundamentals/security/csp/
+        """
+
+        for source in self.get_csp_script_sources(headers) or []:
+            if source.startswith("'nonce-"):
+                return source.partition("'nonce-")[-1][:-1]
+
+    def get_script_with_nonce(self, script, nonce=None):
+        """
+        Given a nonce, get_script_with_nonce returns the injected script text with the nonce.
+
+        If nonce None, get_script_with_nonce returns the script block
+        without attaching a nonce attribute.
+
+        Note:
+        Some responses may specify a nonce inside their Content-Security-Policy,
+        script-src directive.
+        The script injector needs to set the injected script's nonce attribute to
+        open execute permission for the injected script.
+        """
+
+        if nonce:
+            return '<script nonce="{}">{}</script>'.format(nonce, script)
+        return '<script>{}</script>'.format(script)
+
+    def update_csp_script_src(self, headers, sha256):
+        """
+        Update the CSP script directives with appropriate information
+
+        Without this permissions a page with a
+        restrictive CSP will not execute injected scripts.
+
+        Note:
+        For more background information on CSP, please refer to
+        https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/
+        Content-Security-Policy/script-src
+        https://developers.google.com/web/fundamentals/security/csp/
+        """
+
+        sources = self.get_csp_script_sources(headers)
+        add_unsafe = True
+
+        for token in self.get_csp_script_sources(headers):
```

Comment:
We alread have `sources`, and these aren't really tokens. Please update this to `for source in sources`.